### PR TITLE
fix: missing live resources without a health check should not affect application health

### LINKED ISSUE #6283: Exclude configMap sync from the decision to show application healthy/unhealthy.
For showing application healthy/unhealthy ArgoUI considers configMap Sync too. ArgoUI should not decide application healthy/unhealthy based on configMap sync.

### PR DESCRIPTION
**Pull request by alexmt**
Signed-off-by: Alexander Matyushentsev <AMatyushentsev@gmail.com>

Closes #6283 
Depends on https://github.com/argoproj/gitops-engine/pull/285

PR changes how missing status affects aggregated app health status:

* Prioritize Degraded health check over Missing. Some teams don't following 100% gitops e.g. : update manfiests of both QA  and Prod but "sync" Prod only once a week. As a result some Prod app resources are missing most of the time. Prioritizing the Degraded would allow such teams to catch when application Degrades
* Improves handling of Missing status: resources that don't have health check (like ConfigMaps) should not affect health status at all.

### CODE REVIEW

**Comment by jessesuen** [comment]
For anyone wondering the motivation for this fix:

The "Missing" health status has always been a point of confusion for users, and why it relates to health status and not simply sync status. Especially because of the fact that any resource which is "missing", is already covered by OutOfSync. 

The original motivation for a Missing status is for situations for when a Deployment that has not been deployed, does not having the desired replicas running and shouldn't be considered Healthy. Degraded and Progressing are also not appropriate, so Missing was introduced. 

The fix here is that we will only apply "Missing" to objects which *have* health checks and are not yet deployed, so that things like missing ConfigMaps would not affect the aggregate health of the Application (e.g. this for example is important for users using kustomize configmap generators).

**Comment by iam-veeramalla** [comment]
Hi @alexmt @jessesuen 

One of the users reported a similar issue #6241  for an openshift resource `templates`. Template resource also does not have status field but it is an Aggregated API not CRD.  would this PR also handle such scenarios ? or it falls out of scope of this PR?

Thanks :)

**Comment by jessesuen** [comment]
In #6241, it was mentioned that their resource is OutOfSync. Health checks do not affect sync status, they only affect health status so this issue will not help.

**Review by jessesuen — APPROVED** [review]

File: controller/health.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+			continue
+		}
+
+		var healthStatus *health.HealthStatus
+		var err error
+		healthOverrides := lua.ResourceHealthOverrides(resourceOverrides)
+		gvk := schema.GroupVersionKind{Group: res.Group, Version: res.Version, Kind: res.Kind}
+		if res.Live == nil {
+			if _, hasOverride := healthOverrides[lua.GetConfigMapKey(gvk)]; hasOverride || health.GetHealthCheckFunc(gvk) != nil {
+				healthStatus = &health.HealthStatus{Status: health.HealthStatusMissing}
```
[the remark this answers is not in the document]
Comment (jessesuen) [reply controller/health.go:34]: Actually, I have a question. Do we still apply Missing at the resource level or do we simply not let it aggregate up?

My thinking is that `Missing` should still be applied to the resource but we do not let it affect aggregate health.

This is similar to how we allow exceptions for IgnoreExtraneous

[the remark this answers is not in the document]
Comment (alexmt) [reply controller/health.go]: good catch! thank you. Resource still has to be Missing. Fixed and updated unit test accordingly.

File: controller/health.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		if healthStatus != nil {
+			resHealth := appv1.HealthStatus{Status: healthStatus.Status, Message: healthStatus.Message}
+			statuses[i].Health = &resHealth
+
+			// Is health status is missing but resource has not built-in/custom health check then it should not affect parent app health
+			if _, hasOverride := healthOverrides[lua.GetConfigMapKey(gvk)]; healthStatus.Status == health.HealthStatusMissing && !hasOverride && health.GetHealthCheckFunc(gvk) == nil {
+				continue
+			}
+
+			// Missing or Unknown health status of child Argo CD app should not affect parent
```
Comment (tybook) [inline controller/health.go:53]: Hello, I believe this block breaks the use of sync waves with the app-of-apps pattern. Specifically during initial creation of multiple layers of Applications (e.g. "app-of-apps-of-apps"). If sync waves are used on an Application in the intermediate layer of the hierarchy, the initial `Missing` / `Unknown` health statuses of its child Applications don't bubble up, and the sync proceeds to the other intermediate layer Applications even if children Applications haven't been synced and become healthy.

Why does this block need to exist?  Adding back health checks for the app-of-apps pattern is already optional (https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#argocd-app). If a user hasn't added them back, it seems the preceding block on lines 48-51 would already exclude child Application health statuses from affecting the parent's. So, unless I'm missing something, this block on lines 54-56 only has an effect when the user has opted into the app-of-apps health checks, but that effect is to partially break sync waves on Application resources, which is pretty vital for the app-of-apps pattern.

**Pull request merged**